api/src/main/java/org/apache/iceberg/expressions/LiteralSet.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.expressions;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import org.apache.iceberg.util.CharSequenceWrapper;
+
+/**
+ * Represents a set of literal values in an IN or NOT_IN predicate
+ * @param <T> The Java type of the value, which can be wrapped by a {@link Literal}
+ */
+class LiteralSet<T> implements Set<T>, Serializable {
+  private final Set<T> values;
+
+  @SuppressWarnings("unchecked")
+  LiteralSet(Set<Literal<T>> lits) {
+    Preconditions.checkArgument(lits == null || lits.size() > 1,
+        "The input literal set must include more than 1 element.");
+    values = ImmutableSet.<T>builder().addAll(
+        lits.stream().map(
+            lit -> {
+              if (lit instanceof Literals.StringLiteral) {
+                return (T) CharSequenceWrapper.wrap((CharSequence) lit.value());
```

Comment:
If you cast `lit` to `StringLiteral` instead of casting the value, this won't require suppressing unchecked warnings because the `StringLiteral` cast is checked.

Reply:
Update the code, but due to `(T)`, still need suppressing unchecked warning.